### PR TITLE
SITL: add simulated slung payload

### PR DESCRIPTION
**Pull request by rmackay9**
This adds SITL support for a slung payload and is implemented similarly to how the Ship landing ship was implemented.

The payload can be enabled by setting a new SIM_SLUP_ENABLE parameter to 1.  In addition the payload's weight, line length and air resistance (aka drag) can also be adjusted through new parameters

This [Physics Classroom wiki page](https://www.physicsclassroom.com/class/waves/Lesson-0/Pendulum-Motion) was used as a reference to come up with this method for updating the payload and vehicle's movements:

1. update the payload's position, velocity using the previous iterations acceleration
2. check that the payload does not fall below the terrain
3. check if the line is taught and that the payload does not move more than the line length from the vehicle
4. calculate gravity and drag forces on the payload
5. calculate the tension force between the payload and vehicle including force countering gravity, drag and centripetal force
6. update the payload's acceleration using the sum of the above forces
7. update the vehicle's acceleration using the opposite of the above forces

This has been tested in SITL and seems to move in a plausible way.
image
image

To add confidence I've checked that the payload reaches the correct maximum speed according to the above linked pendulum document.  So with drag removed (e.g. SIM_SLUP_DRAG=0) and the vehicle suddenly stopped so as to cause the pendulum to swing out to 38deg the pendulum reached a maximum speed of 11m/s which is very close to the theoretical 11.17m/s.
image

### CODE REVIEW

File: libraries/SITL/SIM_SlungPayload.cpp [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+using namespace SITL;
+
+// SlungPayloadSim parameters
+const AP_Param::GroupInfo SlungPayloadSim::var_info[] = {
+    // @Param: ENABLE
+    // @DisplayName: Slung Payload Sim enable/disable
+    // @Description: Slung Payload Sim enable/disable
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO("ENABLE",  1, SlungPayloadSim,  enable, 0),
```
Comment (tridge) [inline libraries/SITL/SIM_SlungPayload.cpp:40]: mark as an enable param?

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/SITL/SIM_SlungPayload.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/SITL/SIM_SlungPayload.cpp:40]: thanks, done in a follow-up commit

File: libraries/SITL/SIM_Multicopter.cpp [hunk 2]
```diff
@@ -48,6 +48,9 @@ void MultiCopter::calculate_forces(const struct sitl_input &input, Vector3f &rot
 
     add_shove_forces(rot_accel, body_accel);
     add_twist_forces(rot_accel);
+
+    // add forces from slung payload
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/SITL/SIM_Multicopter.cpp:52]: great, thanks.  ifs added

File: libraries/SITL/SIM_SlungPayload.cpp [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+    // @Param: DRAG
+    // @DisplayName: Slung Payload drag coefficient
+    // @Description: Slung Payload drag coefficient.  Higher values increase drag and slow the payload more quickly
+    // @Units: m
+    // @Range: 0 10
+    // @User: Advanced
+    AP_GROUPINFO("DRAG",    4, SlungPayloadSim,  drag_coef, 1),
+
+    // @Param: SYSID
+    // @DisplayName: Slung Payload MAVLink sysmtem ID
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/SITL/SIM_SlungPayload.cpp:67]: spelling fixed, thanks!

File: libraries/SITL/SIM_Aircraft.cpp [hunk 4]
```diff
@@ -1227,6 +1232,19 @@ void Aircraft::add_twist_forces(Vector3f &rot_accel)
     }
 }
 
+// add body-frame force due to slung payload
+void Aircraft::add_slungpayload_forces(Vector3f &body_accel)
+{
+#if AP_SIM_SLUNGPAYLOAD_ENABLED
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/SITL/SIM_Aircraft.cpp:1238]: thanks, moved to outside of fn

**Review by tridge — APPROVED** [review]

**Pull request merged**